python-sdk/nuscenes/predict/models.py

```diff
@@ -0,0 +1,169 @@
+
+def _kinematics_from_tokens(helper: PredictHelper, instance: str, sample: str) -> KinematicsData:
+    """
+    Returns the 2D position, velocity, and acceleration vectors from the given track
+    records, along with the speed, yaw rate, (scalar) acceleration (magnitude), and heading.
+    """
+
+    annotation = helper.get_sample_annotation(instance, sample)
+    x, y, _ = annotation['translation']
+    yaw = quaternion_yaw(Quaternion(annotation['rotation']))
+
+    velocity = helper.get_velocity_for_agent(instance, sample)
+    acceleration = helper.get_acceleration_for_agent(instance, sample)
+    yaw_rate = helper.get_heading_change_rate_for_agent(instance, sample)
+
+    if np.isnan(velocity):
+        velocity = 0.0
+    if np.isnan(acceleration):
+        acceleration = 0.0
+    if np.isnan(yaw_rate):
+        yaw_rate = 0.0
+
+    hx, hy = np.cos(yaw), np.sin(yaw)
+    vx, vy = velocity * hx, velocity * hy
+    ax, ay = acceleration * hx, acceleration * hy
+
+    return x, y, vx, vy, ax, ay, velocity, yaw_rate, acceleration, yaw
+
+
+def _constant_velocity_heading_from_kinematics(kinematics_data: KinematicsData,
+                                               sec_from_now: float,
+                                               sampled_at: int) -> np.ndarray:
+    """
+    Computes a constant velocity baseline for given kinematics data, time window
+    and frequency.
+    """
+    x, y, vx, vy, _, _, _, _, _, _ = kinematics_data
+    preds = []
+    time_step = 1.0 / sampled_at
+    for time in np.arange(time_step, sec_from_now + time_step, time_step):
+        preds.append((x + time * vx, y + time * vy))
+    return np.array(preds)
+
+
+def _constant_acceleration_and_heading(kinematics_data: KinematicsData,
+                                       sec_from_now: float, sampled_at: int) -> np.ndarray:
+    """
+    Computes a baseline prediction for the given time window and frequency, under
+    the assumption that the acceleration and heading are constant.
+    """
+    x, y, vx, vy, ax, ay, _, _, _, _ = kinematics_data
+
+    preds = []
+    time_step = 1.0 / sampled_at
+    for time in np.arange(time_step, sec_from_now + time_step, time_step):
+        half_time_squared = 0.5 * time * time
+        preds.append((x + time * vx + half_time_squared * ax,
+                      y + time * vy + half_time_squared * ay))
+    return np.array(preds)
+
+
+def _constant_speed_and_yaw_rate(kinematics_data: KinematicsData,
+                                 sec_from_now: float, sampled_at: int) -> np.ndarray:
+    """
+    Computes a baseline prediction for the given time window and frequency, under
+    the assumption that the (scalar) speed and yaw rate are constant.
+    """
+    x, y, vx, vy, _, _, speed, yaw_rate, _, yaw = kinematics_data
+
+    preds = []
+    time_step = 1.0 / sampled_at
+    distance_step = time_step * speed
+    yaw_step = time_step * yaw_rate
+    for time in np.arange(time_step, sec_from_now + time_step, time_step):
```

Comment:
Time is not used, can be replaced by `_`.

Reply:
Done!